### PR TITLE
Convert array subscripts with CONVFMT when the element is created

### LINKED ISSUE #547: Single-dimension array subscripts are not converted with CONVFMT when the element is created
`io.jawk.gawk.GawkIT.test_arynasty` was green at [v7.0.01](https://github.com/jawkio/jawk/releases/tag/v7.0.01) and fails on `main` as of 7d0dec5 (#543).

```console
Unexpected output for GAWK arynasty:
  expected:<test[12[.153]] = hi> but was:<test[12[]] = hi>
```

`src/it/resources/gawk/arynasty.awk`:

```awk
BEGIN {
	a = 12.153
	test[a] = "hi"
	CONVFMT = "%.0f"
	a = 5
	for (i in test)
		printf ("test[%s] = %s\n", i, test[i])
}
```

gawk prints `test[12.153] = hi`, matching `arynasty.ok`. Jawk now prints `test[12] = hi`.

### Root cause

An array subscript must be converted to a string with the `CONVFMT` in effect **when the element is created**, and that string is the key from then on. Jawk never converts a single-dimension subscript at all: it stores the raw `Double` as the key and converts lazily on read, so the later `CONVFMT = "%.0f"` retroactively changes the key to `12`.

`APPLY_SUBSEP` is the only opcode that converts a subscript to its string form, and `ArrayIndexAst.populateTuples` emits it only for multi-dimensional subscripts (`src/main/java/io/jawk/frontend/AwkParser.java:4472`):

```java
if (cnt > 1) {
    tuples.applySubsep(cnt);
}
```

For `cnt == 1` the evaluated scalar is pushed and consumed verbatim by the array opcodes.

### Why it surfaced in #543

This is latent and pre-existing, not caused by #543. Before that PR, `printf "%s"` did not apply `CONVFMT`, so the lazily-converted key happened to render as `12.153` and hid the bug. #543 correctly made `%s` honor `CONVFMT` (a documented change in that PR), which exposed it.

### Suggested fix

Always emit the conversion, so every subscript is stringified when it is used:

```java
tuples.applySubsep(cnt);
```

This reportedly needs no AVM change, but it alters the core array path for every script, so it wants a full `mvn verify` and a before/after diff of the failing-test set across all four IT suites. Note that the failing-test comparison must keep the parameterized `[...]` suffix (e.g. `BwkTIT.test[BWK.t t.strnum]`), otherwise per-case changes within a parameterized suite are invisible.

Worth adding a targeted unit test alongside: an element created under one `CONVFMT` must keep its key after `CONVFMT` changes, for both a single and a multi-dimensional subscript.

### PR DESCRIPTION
**Pull request by bertysentry**
Fixes #547.

## Problem

An array subscript must be converted to a string with the `CONVFMT` in effect **when the element is created**, and that string is the key from then on. Jawk never converted a single-dimension subscript: it stored the raw `Double` as the key and converted lazily on read, so a later `CONVFMT` change retroactively altered existing keys. This cost two gawk compatibility tests: `GawkIT.test_arynasty` and `GawkIT.test_ofmta`.

## Fix

- `ArrayIndexAst.populateTuples` now always emits `APPLY_SUBSEP` (previously only when `cnt > 1`), so every subscript is stringified at the point of use — element creation, reads, `delete`, and membership tests.
- The `in` operator converts its key with the `CONVFMT`/`SUBSEP` in effect **when the operator executes**, i.e. after the array operand has been evaluated (it may have side effects), matching gawk's conversion timing and left-to-right operand evaluation order (both gawk-verified). Since the key components sit under the array operand on the stack, a new `APPLY_SUBSEP_UNDER_TOP` opcode (appended at the end of the enum to keep precompiled-tuple ordinals stable) converts them in place.
- Integral subscripts skip the string round-trip: `AssocArray.isIntegralNumberKey()` (next to `normalizeKey()`/`toLongKey()`, whose canonicalization it describes) recognizes an integral `Long`/`Integer`/`Double`, which the array implementations canonicalize to the same `Long` key as its string form — so integer-indexed loops stay fast, `split()`-array interop is unchanged, and injected plain `Map` variables keep exact `Double` key semantics. Every other scalar converts through `jrt.toAwkString()`. Integral values beyond the signed 64-bit range key as their exact digits like gawk: `a[2^63]` creates `9223372036854775808` (previously clamped to `Long.MAX_VALUE`).
- Subscript line numbers are captured where the subscript starts during parsing (shared identifier nodes carry their first occurrence's line, and the lexer reads ahead past line ends), so the "array in a scalar context" error reports the right line for bracketed, grouped, and print-parenthesized subscripts, including multi-line groups.
- `AwkTuples.serialVersionUID` bumped to 4: pre-change precompiled programs lack the conversion tuples and are now rejected with the existing "please recompile" message instead of silently keeping the old behavior.

## Verification

- New script tests in `AssocArrayTest`/`AwkTest` (gawk behavior verified empirically for each semantic case): key fixed at creation (single & multi-dim), creation under non-default `CONVFMT`, `in`/`delete` converting with the current `CONVFMT`, conversion-after-array-operand timing (single & multi-dim), integral subscripts immune to `CONVFMT`, injected plain-`Map` `Double` keys preserved, non-integral `Float` binding keyed at creation, `a[2^63]` exact digits, error line numbers, and a direct unit test of `AssocArray.isIntegralNumberKey()`.
- Full `mvn verify site` passes (surefire, checkstyle, pmd, spotbugs all clean).
- Before/after diff of the failing-test set across all four IT suites (parameterized names preserved), baseline from `main` in a separate worktree:
  - **Fixed**: `io.jawk.gawk.GawkIT.test_arynasty`, `io.jawk.gawk.GawkIT.test_ofmta`
  - **Regressed**: none (151 → 149 failing ITs, stable across all thirteen verify runs of this branch)
- Behavior changes documented under `Unreleased` in `src/site/markdown/behavior-changes.md`.

## Scope note (maintainer decision)

Subscript conversion only distinguishes the number types AWK code itself produces (`Long`/`Integer`/`Double`). Exotic `Number` bindings from JSR-223 or extensions (`Float`, `Short`, `BigInteger`, `BigDecimal`, ...) convert through `jrt.toAwkString()` like any scalar: they do not keep exact-typed keys in injected plain `Map`s, and arbitrary-precision values beyond the double's exactness project through `double`. This keeps the rule lean; the earlier per-type exact-value machinery was removed in e19ed10.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/main/java/io/jawk/backend/AVM.java [hunk 1]
```diff
@@ -3680,6 +3680,27 @@ private void execApplySubsep(CountTuple tuple) {
...
+	 * fixing the key from that point on.
+	 */
+	private Object toSubscriptKey(Object value) {
+		if (value instanceof Long || value instanceof Integer) {
+			return value;
+		}
+		if (value instanceof Double) {
+			Object scalar = JRT.toScalarNumber((Double) value);
+			if (scalar instanceof Long) {
+				return scalar;
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/backend/AVM.java]: Good catch, fixed in 873182d: integral Double subscripts now pass through unchanged (the exact pre-PR value) instead of being converted to Long — the AssocArray implementations canonicalize them at the map layer, and injected plain Maps keep their exact key semantics. Added a test pinning the embedding scenario (plain Map with a Double(1.0) key, Double-preassigned subscript, element access and `in`). Since integral subscripts are now byte-for-byte unaffected, the Unreleased note stays accurate as written. Full `mvn verify` re-run: compatibility results unchanged.

**Comment by bertysentry** [comment]
@codex please review again

File: src/main/java/io/jawk/backend/AVM.java [hunk 2]
```diff
@@ -3680,6 +3680,25 @@ private void execApplySubsep(CountTuple tuple) {
...
+	 * fixing the key from that point on.
+	 */
+	private Object toSubscriptKey(Object value) {
+		if (value instanceof Long || value instanceof Integer) {
+			return value;
+		}
+		if (value instanceof Double && JRT.toScalarNumber((Double) value) instanceof Long) {
+			return value;
+		}
+		return jrt.toAwkString(value);
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/backend/AVM.java]: Fixed in 7410dae: the pass-through now covers any Number whose value is integral and fits a signed 64-bit integer (verified that normalizeExternalVariableValue passes raw Numbers through, so the scenario was real). The exact-type check stays first for Long/Integer/Short/Byte so values near Long.MAX_VALUE are not misjudged through double precision. Test extended to Double, Float, and Short keys. Full `mvn verify` re-run: compatibility results unchanged.

**Comment by bertysentry** [comment]
@codex please review again

File: src/main/java/io/jawk/backend/AVM.java [hunk 3]
```diff
@@ -3680,6 +3680,26 @@ private void execApplySubsep(CountTuple tuple) {
...
+	 * string round-trip keeps integer-indexed loops fast, and injected plain
+	 * {@code Map} variables keep their exact key semantics. Every other
+	 * scalar is converted to a string with the CONVFMT currently in effect,
+	 * fixing the key from that point on.
+	 */
+	private Object toSubscriptKey(Object value) {
+		if (value instanceof Long || value instanceof Integer || value instanceof Short || value instanceof Byte) {
+			return value;
+		}
+		if (value instanceof Number && JRT.toScalarNumber(((Number) value).doubleValue()) instanceof Long) {
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/backend/AVM.java]: Fixed in c5fc66e by inverting the rule so no Number is ever projected through double to decide its fate: among numbers, only a non-integral Double is CONVFMT-converted (Double is the only floating-point type AWK code can produce, and it is exactly where the bug was). Every other Number — Float, Short, BigInteger, anything JSR-223 can bind — passes through unchanged, i.e. exact pre-PR behavior for injected plain Maps, so boundary values like BigInteger.valueOf(Long.MAX_VALUE) are untouched by construction. Test extended with that boundary key. Full `mvn verify` re-run: compatibility results unchanged.

**Comment by bertysentry** [comment]
@codex please review again

File: src/main/java/io/jawk/backend/AVM.java [hunk 4]
```diff
@@ -3680,6 +3680,31 @@ private void execApplySubsep(CountTuple tuple) {
...
+	 * integer-indexed loops fast), and the remaining {@code Number} types
+	 * can only be bound by Java callers, whose injected plain {@code Map}
+	 * variables keep their exact key semantics.
+	 */
+	private Object toSubscriptKey(Object value) {
+		if (value instanceof Long || value instanceof Integer) {
+			return value;
+		}
+		if (value instanceof Double) {
+			return JRT.toScalarNumber(((Double) value).doubleValue()) instanceof Long ? value : jrt.toAwkString(value);
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/backend/AVM.java]: This one is deliberate, and the suggested pass-through would preserve a bug. Verified against real gawk: `a[2^63]` keys as the exact digit string `9223372036854775808` regardless of CONVFMT — exactly what this branch produces. Before this PR, Jawk clamped such keys to Long.MAX_VALUE (printing `9223372036854775807`, off by one, and colliding every out-of-range integral subscript on a single key), so preserving the raw Double for the injected-plain-Map edge would keep Jawk wrong on the AWK side, against the project's gawk-compat policy. Pinned in bb0ff71 with a gawk-verified test, and the Unreleased note now covers the boundary explicitly (in-range integer subscripts unaffected; beyond-range integral subscripts key as exact digits, previously clamped).

**Comment by bertysentry** [comment]
@codex please review again

File: src/main/java/io/jawk/frontend/AwkParser.java [hunk 5]
```diff
@@ -4217,6 +4217,11 @@ public int populateTuples(AwkTuples tuples) {
 			}
 
 			getAst1().populateTuples(tuples);
+			if (!(getAst1() instanceof ArrayIndexAst)) {
+				// A bare scalar key ("x in a") is converted with the current
+				// CONVFMT, just like a bracketed subscript would be.
+				tuples.applySubsep(1);
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/frontend/AwkParser.java]: Fixed in 4f45273, verified against real gawk on both points: the key converts after the array operand runs (your scenario prints 0), and the key expression still evaluates before the array operand. Single-key forms now emit raw key → array operand → swap / APPLY_SUBSEP / swap, so the conversion uses the CONVFMT in effect when the operator executes while preserving gawk's operand evaluation order. Gawk-verified test added. Full `mvn verify` re-run: compatibility results unchanged.

**Comment by bertysentry** [comment]
@codex please review again

File: src/main/java/io/jawk/frontend/AwkParser.java [hunk 6]
```diff
@@ -4456,7 +4474,11 @@ public int populateTuples(AwkTuples tuples) {
 	private final class ArrayIndexAst extends ScalarExpressionAst {
 
 		private ArrayIndexAst(AST exprAst, AST next) {
-			super(exprAst, next);
+			// Anchor to the first subscript expression's line: by the time this
+			// node is built the lexer may have read past the end of the line,
+			// and the APPLY_SUBSEP tuple must report errors where the subscript
+			// starts.
+			super(exprAst.getLineNo(), exprAst, next);
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/frontend/AwkParser.java]: Fixed in 1460b22: ARRAY_INDEX captures the source line where the subscript starts, before parsing the expression, and every ArrayIndexAst construction site now passes an explicit line — no more deriving it from the shared identifier node. Your exact scenario is pinned by a new line-number test.

File: src/site/markdown/behavior-changes.md [hunk 7]
```diff
@@ -20,6 +20,17 @@ released version automatically via .github/scripts/stamp-behavior-changes.sh.
 
 ## Unreleased
 
+- Array subscripts are now converted to their string key with the `CONVFMT` in effect at the
+  moment the subscript is used, as POSIX requires and gawk does. Previously a single-dimension
+  numeric subscript was stored as a number and converted lazily, so a later `CONVFMT` change
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/site/markdown/behavior-changes.md:25]: Done in 1460b22: the Unreleased entry now records that the array-as-scalar-subscript error fires when the subscript is converted and is reported on the line where the subscript starts (previously raised by the consuming operation on the enclosing array reference's line).

**Comment by bertysentry** [comment]
@codex please review again

File: src/main/java/io/jawk/frontend/AwkParser.java [hunk 8]
```diff
@@ -4216,8 +4221,26 @@ public int populateTuples(AwkTuples tuples) {
...
+				rawKeyAst.populateTuples(tuples);
+				populateArrayOperandTuples(getAst2(), tuples, false, "Expecting an array for rhs of IN. Got a scalar.");
+				tuples.swap();
+				tuples.applySubsep(1);
+				tuples.swap();
+			} else {
+				// A multi-dimensional key joins its components with SUBSEP as it
+				// is built, before the array operand is evaluated.
+				keyAst.populateTuples(tuples);
+				populateArrayOperandTuples(getAst2(), tuples, false, "Expecting an array for rhs of IN. Got a scalar.");
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/frontend/AwkParser.java:4248]: Fixed in a425b46, gawk-verified: the multi-dimensional key now joins and converts after the array operand runs (your scenario prints 0), with components still evaluated left-to-right first. A new APPLY_SUBSEP_UNDER_TOP opcode (appended at the enum end to keep precompiled-tuple ordinals stable) converts the components under the array operand in place; both in forms share this emission now, which also removed the swap pair. Pinned by a gawk-verified test.

File: src/main/java/io/jawk/backend/AVM.java [hunk 9]
```diff
@@ -3680,6 +3680,31 @@ private void execApplySubsep(CountTuple tuple) {
...
+		}
+		if (value instanceof Number) {
+			return value;
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/backend/AVM.java]: Fixed in a425b46: toSubscriptKey recognizes integral numbers by type for the integer classes (Long/Integer/Short/Byte/BigInteger — never projected through double, so the earlier boundary findings stay fixed) and by value for the floating-point classes, so a non-integral Float or BigDecimal from an extension or binding converts to its CONVFMT string at creation like a Double. Pinned with a Float(12.153f) test: key stays 12.153 after CONVFMT changes, membership under %.0f is 0.

File: src/main/java/io/jawk/frontend/AwkParser.java [hunk 10]
```diff
@@ -1613,10 +1613,11 @@ private AST ASSIGNMENT_EXPRESSION(
 			lexer();
 			optNewline();
 			AST rest = ASSIGNMENT_EXPRESSION(null, allowComparison, allowInKeyword, true);
+			int lineNo = currentSourceLineNumber();
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/frontend/AwkParser.java]: Fixed in a425b46: ASSIGNMENT_EXPRESSION captures the group's line before parsing begins, and the in operator emits its key conversion under the subscript's start line, so a multi-line grouped key reports where the subscript starts. Pinned by a line-number test with the group spanning two lines.

**Comment by bertysentry** [comment]
@codex please review again

File: src/main/java/io/jawk/backend/AVM.java [hunk 11]
```diff
@@ -3680,6 +3690,35 @@ private void execApplySubsep(CountTuple tuple) {
...
+	private Object toSubscriptKey(Object value) {
+		if (value instanceof Long
+				|| value instanceof Integer
+				|| value instanceof Short
+				|| value instanceof Byte
+				|| value instanceof BigInteger) {
+			return value;
+		}
+		if (value instanceof Number && JRT.toScalarNumber(((Number) value).doubleValue()) instanceof Long) {
+			return value;
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/backend/AVM.java]: Fixed in f261e6f: non-Double Numbers are now tested against their exact decimal value (integral scale + signed-64-bit bounds via BigDecimal), so BigDecimal.valueOf(Long.MAX_VALUE) — and any other Number type, AtomicLong included — passes through when its value is a signed-64-bit integer. No double projection remains anywhere in the rule; Double keeps its own exact fast path. Pinned by extending the injected-map test with the BigDecimal boundary key.

File: src/main/java/io/jawk/frontend/AwkParser.java [hunk 12]
```diff
@@ -2447,7 +2456,7 @@ private AST parseParenthesizedPrintArguments() throws IOException {
 	// it turned out to be, when the parenthesized group is followed by "in"
 	private AST toMultidimIndex(FunctionCallParamListAst list) {
 		AST rest = list.getAst2() == null ? null : toMultidimIndex((FunctionCallParamListAst) list.getAst2());
-		return new ArrayIndexAst(list.getAst1(), rest);
+		return new ArrayIndexAst(list.getLineNo(), list.getAst1(), rest);
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/frontend/AwkParser.java:2464]: Fixed in f261e6f: EXPRESSION_LIST captures the line where its first expression starts and passes it to FunctionCallParamListAst, so a print group rebuilt into a multi-dimensional in key reports the scalar-context error on the subscript's start line even when the group spans lines. Pinned by a line-number test on the print (a, / 1) in b shape.

**Comment by bertysentry** [comment]
@codex please review again

File: src/main/java/io/jawk/backend/AVM.java [hunk 13]
```diff
@@ -3680,6 +3690,48 @@ private void execApplySubsep(CountTuple tuple) {
...
+				BigDecimal decimal = new BigDecimal(value.toString());
+				if (decimal.stripTrailingZeros().scale() <= 0
+						&& decimal.compareTo(MIN_LONG_SUBSCRIPT) >= 0
+						&& decimal.compareTo(MAX_LONG_SUBSCRIPT) <= 0) {
+					return value;
+				}
+			} catch (NumberFormatException e) { // NOPMD - EmptyCatchBlock: NaN/infinity converts as a string
+			}
+		}
+		return jrt.toAwkString(value);
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/backend/AVM.java]: Fixed in 2aaec80: an integral value beyond the signed 64-bit range now keys as its exact digits straight from the parsed BigDecimal — no double round-trip — matching the exact-digits behavior of the out-of-range integral Double path. Pinned with a test storing 2^63 and 2^63+1 BigInteger subscripts: length(a) is 2 and both exact-digit keys are reachable.

**Comment by bertysentry** [comment]
@codex please review again

File: src/main/java/io/jawk/backend/AVM.java [hunk 14]
```diff
@@ -3663,7 +3673,7 @@ private void execApplySubsep(CountTuple tuple) {
 		if (count == 1) {
 			Object value = pop();
 			checkScalar(value);
-			push(jrt.toAwkString(value));
+			push(toSubscriptKey(value));
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/backend/AVM.java]: Fixed in c3fe50d: each multi-dimensional component now converts through the same rules as a single subscript (via a shared helper) before the SUBSEP join, so beyond-range integral components keep their exact digits. Pinned by extending the arbitrary-precision test with a[n1, 0] / a[n2, 0] — length is 2 and the exact-digit joined key is reachable.

**Comment by bertysentry** [comment]
@codex please review again

File: src/main/java/io/jawk/frontend/AwkParser.java [hunk 15]
```diff
@@ -4455,25 +4485,38 @@ public int populateTuples(AwkTuples tuples) {
...
 		}
 
 		@Override
 		public int populateTuples(AwkTuples tuples) {
 			pushSourceLineNumber(tuples);
+			int cnt = populateComponentTuples(tuples);
+			// Convert the subscript to its string form now, with the CONVFMT in
+			// effect at this point of the execution: a later CONVFMT change must
+			// not retroactively alter existing keys.
+			tuples.applySubsep(cnt);
```
[the remark this answers is not in the document]
Comment (bertysentry) [reply src/main/java/io/jawk/frontend/AwkParser.java:4501]: Fixed in e21bba0: AwkTuples.serialVersionUID bumped to 4, so a pre-change precompiled program loaded with -L is rejected through the existing InvalidClassException path with the "not compatible with this version, please recompile" message instead of silently keeping the lazy-key behavior.

**Comment by bertysentry** [comment]
@codex please review again

**Comment by bertysentry** [comment]
@codex please review again
